completion_aggregator/tasks.py

```diff
@@ -0,0 +1,188 @@
+"""
+Asynchronous tasks.
+"""
+
+from __future__ import absolute_import, division, print_function, unicode_literals
+
+from datetime import datetime
+
+import pytz
+from celery import shared_task
+from xblock.completable import XBlockCompletionMode
+from xblock.core import XBlock
+
+from django.conf import settings
+
+from .models import Aggregator
+
+OLD_DATETIME = pytz.utc.localize(datetime(1900, 1, 1, 0, 0, 0))
+
+
+@shared_task
+def update_aggregators(user, course_key, block_keys=frozenset()):  # pylint: disable=unused-argument
+    """
+    Update aggregators for the specified course.
+
+    Takes a collection of block_keys that have been updated, to enable
+    future optimizations in how aggregators are recalculated.
+    """
+    from xmodule.modulestore.django import modulestore   # pylint: disable=import-error
+
+    updater = AggregationUpdater(user, course_key, modulestore())
+    updater.update(block_keys)
+
+
+class AggregationUpdater(object):
+    """
+    Class to update aggregators for a given course and user.
+    """
+
+    def __init__(self, user, course_key, modulestore):
+        """
+        Create an aggregation updater for the given user and course.
+
+        Also takes a modulestore instance.
+        """
+        self.user = user
+        self.course_key = course_key
+        with modulestore.bulk_operations(self.course_key):
+            self.course_block_key = self.init_course_block_key(modulestore, self.course_key)
+            self.course_blocks = self.init_course_blocks(self.user, self.course_block_key)
+        self.aggregators = {
+            aggregator.block_key: aggregator for aggregator in Aggregator.objects.filter(
+                user=self.user,
+                course_key=self.course_key,
+            )
+        }
+        self.block_completions = {completion.block_key: completion for completion in self._get_block_completions()}
+
+    def update(self, changed_blocks=frozenset()):
+        """
+        Update the aggregators for the course.
+
+        Takes a set of completable blocks that have been recently updated to
+        inform how to perform the update.  Currently no optimizations are
+        performed based on this information, but in the future they may help
+        cut down on the amount of work performed.
```

Comment:
Is the idea that, since you have the BlockStructure for the course, you can just search the structure for any node with a corresponding aggregator that has one of the `changed_blocks` in its subtree?

Reply:
Yeah, that's pretty much it.  It seems that once you've traversed the course graph, summing the completions is pretty near free (computers are really good at addition), but if we can decide *before* traversing the tree that we trust a given aggregator's value, we can skip the traversal altogether.  Whether that's by calculating the descendents of an aggregator block or the ancestors of the completable blocks, I'm not sure.